_blogposts/2023-05-17-enhancing-the-ergonomics-of-records.mdx

```diff
@@ -0,0 +1,140 @@
+---
+author: rescript-team
+date: "2023-05-17"
+title: Enhancing the ergonomics of records
+badge: roadmap
+description: |
+  A tour of new capabilities coming in ReScript v11
+---
+
+[Records](https://rescript-lang.org/docs/manual/latest/record) are a fundamental part of ReScript, representing an efficient and expressive way to model structured data. Beyond offering a clear and concise definition of complex data structures, records bring numerous advantages such as strong type checking, immutability by default, great error messages, and support for exhaustive pattern matching. This allows you to write robust and maintainable code, while the compiler ensures you've handled all possible cases when dissecting a record.
+
+However, there's always room for improvement. In ReScript v10 we added support for [optional record fields](https://rescript-lang.org/docs/manual/latest/record#optional-record-fields) to make constructing records with many optional fields easier and more ergonomic. Now we're focusing on how we can improve the workflow around defining records.
+
+In ReScript v11, we're delighted to introduce two new features that will significantly boost your record manipulation capabilities: Record Type Spreads and Record Type Coercion. Let's delve into what these new features offer and how they can make working with records more ergonomic.
+
+## Type Spreads
+
+Before type spreads, creating a new record that was similar or extended from another record required spelling out every single field in the new definition. This was often tedious, error-prone and made code harder to maintain, especially when working with records with many fields.
+
+In ReScript v11, you can now spread one or more record types into a new record type. It looks like this:
+
+```rescript
+type a = {
+  id: string,
+  name: string,
+}
+
+type b = {
+  age: int
+}
+
+type c = {
+  ...a,
+  ...b,
+  active: bool
+}
+```
+
+`type c` will now be:
+
+```rescript
+type c = {
+  id: string,
+  name: string,
+  age: int,
+  active: bool,
+}
+```
+
+Keeping it as straightforward as possible, spreads are essentially a 'copy-paste' operation for fields from one or more records to another, inlining the fields from the spread records into the new record.
```

Comment:
What happens if I spread into a record that overrides an attribute?

```res
module Name = {
  type t
}

type a = {
  name: string
}

type b = {
  age: int
}

type c = {
  ...a,
  ...b,
  name: Name.t
}
```

does this error? \cc @cristianoc

Reply:
Yes, this errors right now. Cannot have overlapping properties. I omitted saying that in the post intentionally btw, just because it's a detail.

Reply:
Added a quick sentence about it just to be safe.